### PR TITLE
added cmake support

### PR DESCRIPTION
**Pull request by berndpfrommer**
added support for building under cmake

### CODE REVIEW

File: cmake/vilibConfig.cmake.in [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+# find_dependency has the same syntax as find_package.
+# List all dependencies here so they are exposed to
+# the importing package through the vilibConfig.cmake
+
+# CUDA is not a dependency because it is natively
+# provided by cmake as language support?
+# Either way, there is no cudaConfig right now.
+
+# find_dependency(cuda REQUIRED)
```
Comment (foehnx) [inline cmake/vilibConfig.cmake.in]: remove unnecessary comments

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,157 @@
+# ----------------------------------------------------------------------
+# 2020 Bernd Pfrommer bernd.pfrommer@gmail.com
+#
```
Comment (foehnx) [inline CMakeLists.txt]: Please include our copyright header as in https://github.com/uzh-rpg/vilib/blob/master/visual_lib/Makefile#L1

In the files you edited, if you whish (and take responsibility) you can add your name to the header in a section below the part that says
```
# Copyright (c) 2019-2020 Balazs Nagy,
# Robotics and Perception Group, University of Zurich
```

File: cmake/vilibConfig.cmake.in [hunk 3]
```diff
@@ -0,0 +1,17 @@
+include(CMakeFindDependencyMacro)
```
Comment (foehnx) [inline cmake/vilibConfig.cmake.in:31]: Please include our copyright header as above.

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+  visual_lib/src/feature_detection/detector_base_gpu.cpp
+  visual_lib/src/feature_detection/detector_benchmark.cpp
+  visual_lib/src/feature_detection/detector_base.cpp
+  visual_lib/src/timergpu.cpp
+  visual_lib/src/feature_tracker/feature_tracker_cuda_tools.cu
+  visual_lib/src/preprocess/pyramid_gpu.cu
+  visual_lib/src/feature_detection/fast/fast_gpu_cuda_tools.cu
+  visual_lib/src/feature_detection/detector_base_gpu_cuda_tools.cu
+  )
+#  visual_lib/src/cuda_common.cpp
```
Comment (foehnx) [inline CMakeLists.txt]: why not getting all source files by
```
file(GLOB_RECURSE SOURCES
  visual_lib/src/*.cpp
  visual_lib/src/*.cu
)
```
and probably remove the ones you don't want with
```
set(EXCLUDE_SOURCES "visual_lib/src/cuda_common.cpp")
list(REMOVE_ITEM SOURCES ${EXCLUDE_SOURCES})
```

**Review by foehnx — CHANGES_REQUESTED** [review]
Thanks a lot, this is nice.
I quickly tested it and seems to work with CMake (3.16.3), gcc (9.3.0) and nvcc (11.2.67).

It would be nice if you could add quick building/installing instructions to the README, and how it can be used/included this way in other projects.

Have you run a runtime comparison to check if the compiler flags differ and/or have impact with respect to the makefile-built version?

**Comment by berndpfrommer** [comment]
Thanks for reviewing. 
I have not done any runtime performance testing, nor verified that the compiler flags are identical. I added the cmake support to use vilib from within a ROS2 wrapper package, and was very happy with the performance I got there, so I stopped at that.

Traveling right now with a laptop that has no nvidia card. Will do the testing and make the changes you requested when returning in about 2 weeks.

**Comment by foehnx** [comment]
awesome, thank you.

If your ROS2 project is open source, feel free to mention it in the README and link to it, if you like.

File: cmake/vilibConfig.cmake.in (hunk 1 above)
Comment (berndpfrommer) [reply cmake/vilibConfig.cmake.in]: done

File: CMakeLists.txt (hunk 2 above)
Comment (berndpfrommer) [reply CMakeLists.txt]: done

File: cmake/vilibConfig.cmake.in (hunk 3 above)
Comment (berndpfrommer) [reply cmake/vilibConfig.cmake.in:31]: done

File: CMakeLists.txt (hunk 4 above)
Comment (berndpfrommer) [reply CMakeLists.txt]: not done because I didn't know of this handy feature :)
done.

**Comment by berndpfrommer** [comment]
So far I had only used the cmake package in my ROS2 package and it worked fine there. When creating a stand-alone package I discovered some transitive dependencies were not propagated correctly.
This has now been fixed. Will squash the commits and push again.

Here is a performance comparison between original and cmake library tests
[vilib_orig.txt](https://github.com/uzh-rpg/vilib/files/5852263/vilib_orig.txt)
[vilib_cmake.txt](https://github.com/uzh-rpg/vilib/files/5852267/vilib_cmake.txt)

**Comment by berndpfrommer** [comment]
I tested once more on Ubuntu 18.04 (it compiles, but not run tested) and 20.04 (compiles and run tested).
A section with instructions on how to use has been added to the README file.
Please have a look at it again.

**Review by foehnx — APPROVED** [review]
Looks good to me!
Thanks a lot for your contribution!

**Pull request merged**